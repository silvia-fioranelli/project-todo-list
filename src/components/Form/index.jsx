import { func } from 'prop-types';
import { Component } from 'react';

class Form extends Component {
  constructor(props) {
    super(props);
    this.state = {
      inputText: '',
    };
  }

  handleChange = (event) => {
    this.setState({
      inputText: event.target.value,
    });
  };

  handleSubmit = (event) => {
    event.preventDefault();

    const { onSubmit } = this.props;
    const { inputText } = this.state;
    
    if (!inputText) return;
    
    onSubmit(inputText);
    this.setState({
      inputText: '',
    });
  };

  render() {
    const { inputText } = this.state;

    return (
      <form>
        <input
          label="Add todo"
          value={inputText}
          type="text"
          onChange={this.handleChange}
        />
        <button onClick={this.handleSubmit}>Add</button>
      </form>
    );
  }
}

Form.propTypes = {
  onSubmit: func.isRequired,
};

export default Form;
