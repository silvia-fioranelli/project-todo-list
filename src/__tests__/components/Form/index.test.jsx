import { shallow } from 'enzyme';
import Form from '../../../components/Form';

describe('Form', () => {
  const defaultProps = {
    onSubmit: jest.fn(),
  };

  const render = (props) => shallow(<Form {...defaultProps} {...props} />);

  describe('rendering', () => {
    test('should render text form', () => {
      const wrapper = render();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('behaviour', () => {
    test('should handle onChange event', () => {
      const wrapper = render();
      wrapper.instance().handleChange = jest.fn();
      // needed by enzyme to register mock callback
      wrapper.instance().forceUpdate(); 
      
      wrapper.find('input').simulate('change');
      expect(wrapper.instance().handleChange).toBeCalledTimes(1);
    });

    test('should handle onClick event', () => {
      const wrapper = render();
      wrapper.instance().handleSubmit = jest.fn();
      // needed by enzyme to register mock callback
      wrapper.instance().forceUpdate(); 
      
      wrapper.find('button').simulate('click');
      expect(wrapper.instance().handleSubmit).toBeCalledTimes(1);
    });
  });

  describe('callbacks', () => {
    test('handleChange should set state', () => {
      const instance = render().instance();
      const event = { target: { value: 'mock' } };
      instance.handleChange(event);
      expect(instance.state.inputText).toEqual('mock');
    });

    test('handleSubmit should not set todo', () => {
      const props = {
        onSubmit: jest.fn(),
      };
      const event = {
        preventDefault: jest.fn(),
      };
      const instance = render(props).instance();
      instance.setState({ inputText: '' });
      instance.handleSubmit(event);

      expect(event.preventDefault).toBeCalledTimes(1);
      expect(props.onSubmit).toBeCalledTimes(0);
    });

    test('handleSubmit should set todo', () => {
      const props = {
        onSubmit: jest.fn(),
      };
      const event = {
        preventDefault: jest.fn(),
      };
      const instance = render(props).instance();
      instance.setState({ inputText: 'mock' });
      instance.handleSubmit(event);

      expect(event.preventDefault).toBeCalledTimes(1);
      expect(props.onSubmit).toBeCalledTimes(1);
      expect(props.onSubmit).toBeCalledWith('mock');
      expect(instance.state.inputText).toEqual('');
    });
  });
});
