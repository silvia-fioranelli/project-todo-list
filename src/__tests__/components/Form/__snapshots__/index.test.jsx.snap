// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form rendering should render text form 1`] = `
<form>
  <div
    className="input-group mb-3"
  >
    <input
      onChange={[Function]}
      placeholder="write a reminder..."
      type="text"
      value=""
    />
    <div
      class="input-group-append"
    >
      <button
        className="btn btn-primary btn-sm"
        onClick={[Function]}
      >
        Add
      </button>
    </div>
  </div>
</form>
`;
